executor/merge_join.go

```diff
@@ -0,0 +1,419 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/plan"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util/types"
+	"sync/atomic"
+)
+
+// MergeJoinExec implements the merge join algorithm.
+// This operator assumes that two iterators of both sides
+// will provide required order on join condition:
+// 1. For equal-join, one of the join key from each side
+// matches the order given.
+// 2. For other cases its preferred not to use SMJ and operator
+// will throw error.
+type MergeJoinExec struct {
+	// Left is always the driver side
+	ctx           context.Context
+	stmtCtx       *variable.StatementContext
+	leftJoinKeys  []*expression.Column
+	rightJoinKeys []*expression.Column
+	prepared      bool
+	leftFilter    expression.Expression
+	rightFilter   expression.Expression
+	otherFilter   expression.Expression
+	schema        *expression.Schema
+	preserveLeft  bool
+	preserveRight bool
+	cursor        int
+	defaultValues []types.Datum
+	// Default for both side in case full join
+	defaultLeftRow  *Row
+	defaultRightRow *Row
+	outputBuf       []*Row
+	finished        atomic.Value
+
+	leftRowBlock  rowBlockIterator
+	rightRowBlock rowBlockIterator
+	leftRows      []*Row
+	rightRows     []*Row
+	desc          bool
+}
+
+const rowBufferSize = 4096
+
+// NewMergeJoinExec is the Constructor for mergeJoinExec
+func NewMergeJoinExec(
```

Comment:
I think a struct is better, like this:
https://github.com/pingcap/tidb/blob/master/executor/analyze.go#L103